word2vec_decade_experiment/03_word_decade_figure_generator_2000_2020.py

```diff
@@ -0,0 +1,315 @@
+# ---
+# jupyter:
+#   jupytext:
+#     formats: ipynb,py
+#     text_representation:
+#       extension: .py
+#       format_name: light
+#       format_version: '1.5'
+#       jupytext_version: 1.9.1+dev
+#   kernelspec:
+#     display_name: Python [conda env:biovectors]
+#     language: python
+#     name: conda-env-biovectors-py
+# ---
+
+# # Generate Figures for Word2Vec Year Models
+
+# The goal of this notebook is to observe how words are shifting through time since 2000. The caveat here is that words have to been present within all time periods in order to be present for this task. Ideally words to be examined so far are: 'expression', 'are', 'interleukin-18', '95%ci', and 'p53'.
+
+# +
+# %load_ext autoreload
+# %autoreload 2
+
+from pathlib import Path
+import pickle
+
+from msp_tsne import MultiscaleParametricTSNE
+import numpy as np
+import pandas as pd
+import plotnine as p9
+import tensorflow as tf
+from tensorflow import keras
+import tqdm
+
+from biovectors_modules.plot_helper import (
+    plot_local_global_distances,
+    plot_token_timeline,
+    plot_wordcloud_neighbors,
+    plot_wordcloud_neighbors_gif,
+)
+from biovectors_modules.word2vec_analysis_helper import (
+    generate_timeline,
+    get_neighbors,
+    project_token_timeline,
+    window,
+)
+# -
+
+# # Load Models to Observe Changes
+
+# ## Load Aligned Word Vectors
+
+aligned_models = pickle.load(open("output/aligned_word_vectors.pkl", "rb"))
+
+year_comparison_dict = {
+    "_".join(comparison_file.stem.split("_")[0:2]): (
+        pd.read_csv(str(comparison_file), sep="\t")
+    )
+    for comparison_file in (list(Path("output/year_distances").rglob("*tsv")))
+}
+list(year_comparison_dict.keys())[0:3]
+
+year_comparison_dict["2007_2008"].sort_values("global_dist")
+
+# ## Train TSNE Model to Project Time Shifts into Two Dimensional Space
+
+# The goal here is to train a TSNE model that projects all words from 2000 to 2020 into a two dimensional space. Allows one to visually track how a word vector is shifting through time.
+
+word_models_stacked = np.vstack(list(aligned_models.values())[:-1])
+file_name = "output/2000_2020_model.h5"
+
+if not Path(file_name).exists():
+    tf.random.set_random_seed(100)
+    np.random.seed(100)
+    model = MultiscaleParametricTSNE(n_iter=300)
+    model.fit(word_models_stacked)
+    keras.models.save_model(model._model, file_name)
+else:
+    model = MultiscaleParametricTSNE(n_iter=300)
+    model._build_model(300, 2)
+    model._model.load_weights(file_name)
+
+# # Observe Diachronic Vector Changes
+
+# Global distance measures how far a word has moved within semantic space. This measure captures how words change globally across time periods. The greater the distance the more semantic change a word has been subjected towards.
+# The word clouds depict the neighbors for each word vector. The size for each token  appears to be different but size doesn't matter in this case. Each word has equal weighting.
```

Comment:
I have a basic question regarding global measures: from what I recall from your presentation, you train a word2vec model on each time period. So here for the global distance, are you just computing the distance (Euclidean? Or... I think you mentioned cosine distance) between vectors of the same word? If this is the case, what's the meaning of this global distance between different representations? I'm not an expert on this, so this might make no sense at all.

Reply:
>I have a basic question regarding global measures: from what I recall from your presentation, you train a word2vec model on each time period. So here for the global distance, are you just computing the distance (Euclidean? Or... I think you mentioned cosine distance) between vectors of the same word? 

Yes exactly. I'm measuring the cosine distance between both vectors after being aligned to the most recent year.

>If this is the case, what's the meaning of this global distance between different representations?

This means the token vectors are more dissimilar (up in distantance) towards one another indicating there is a shift in semantic meaning (e.g. concepts associated with the given token).